File: java/src/main/java/org/xllgl2017/Response.java
package org.xllgl2017;

import com.google.gson.Gson;
import com.google.gson.JsonElement;
import com.google.gson.reflect.TypeToken;
import com.sun.jna.Pointer;
import com.sun.jna.ptr.LongByReference;
import com.sun.jna.ptr.PointerByReference;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.NoSuchElementException;

import static org.xllgl2017.ReqrioLibrary.REQRIO;

public class Response {
    private final int statusCode;
    private final Headers headers = new Headers();
    private final byte[] body;
    private final int sid;
    private final Pointer req;


    public Response(Pointer raw, Pointer req) throws Exception {
        try {
            this.req = req;
            // status code
            PointerByReference err = new PointerByReference();
            this.statusCode = REQRIO.Response_status_code(raw, err);
            util.check_err_pointer(err);

            //header
            Pointer key_ptr = REQRIO.Response_header_keys(raw, err);
            util.check_err_pointer(err);
            String[] keys = key_ptr.getString(0).split(",,,,");
            REQRIO.char_free(key_ptr);
            for (String key : keys) {
                Pointer ptr = REQRIO.Response_get_header(raw, key, err);
                util.check_err_pointer(err);
                this.headers.addHeader(key, ptr.getString(0));
                REQRIO.char_free(ptr);
            }

            //cookies
            Pointer ptr = REQRIO.Response_cookies(raw, err);
            util.check_err_pointer(err);
            String cookie_str = ptr.getString(0);
            REQRIO.char_free(ptr);
            Gson gson = new Gson();
            ArrayList<Cookie> cookies = gson.fromJson(cookie_str, new TypeToken<ArrayList<Cookie>>() {
            }.getType());
            this.headers.setCookies(cookies);

            //body
            LongByReference len = new LongByReference();
            Pointer body_ptr = REQRIO.Response_bytes(raw, len, err);
            util.check_err_pointer(err);
            this.body = body_ptr.getByteArray(0, (int) len.getValue());
            this.sid = REQRIO.Response_sid(raw, err);
            util.check_err_pointer(err);
        } finally {
            REQRIO.Response_drop(raw);
        }
    }

    public int statusCode() {
        return this.statusCode;

    }

    public String getHeader(String name) {
        return this.headers.get(name);
    }

    public String location() {
        return this.getHeader("location");
    }

    public byte[] bytes() {
        return this.body;
    }

    public String text() {
        return new String(this.bytes());
    }

    public JsonElement json() {
        Gson gson = new Gson();
        return gson.fromJson(this.text(), JsonElement.class);
    }

    public List<Cookie> getCookies() {
        return headers.getCookies();
    }

    public Iterable<byte[]> chunks() {
        return () -> new Iterator<>() {
            boolean hashNext = true;
            byte[] chunk;

            @Override
            public boolean hasNext() {
                LongByReference len = new LongByReference();
                PointerByReference err = new PointerByReference();
                Pointer ptr = REQRIO.ScReq_recv_stream(req, sid, len, err);
                try {
                    util.check_err_pointer(err);
                } catch (Exception e) {
                    throw new RuntimeException(e);
                }
                if (ptr == Pointer.NULL) {
                    hashNext = false;
                } else {
                    this.chunk = ptr.getByteArray(0, (int) len.getValue());
                }
                return hashNext;
            }

            @Override
            public byte[] next() {
                if (!hashNext){
                    throw new NoSuchElementException();
                }
                return this.chunk;
            }
        };
    }

    public int getSid() {
        return sid;
    }
}

File: qt/Response.cpp
//
// Created by XLX on 2026/1/1.
//

#include "Response.h"

#include <qdebug.h>
#include <QJsonArray>
#include <QJsonDocument>
#include <QJsonObject>

#include "Body.h"
#include "util.h"


Response::Response(bindings::Response *ptr, bindings::ScReq *req, QObject *parent) : QObject(parent) {
    this->raw_ptr = ptr;
    this->req_ptr = req;
}

int Response::statusCode() const {
    char *err = nullptr;
    const int code = bindings::Response_status_code(this->raw_ptr, &err);
    util::check_err(err);
    return code;
}

QByteArray Response::bytes() const {
    char *err = nullptr;
    size_t len = 0;
    uint8_t *ptr = bindings::Response_bytes(this->raw_ptr, &len, &err);
    util::check_err(err);
    QByteArray bytes = QByteArray::fromRawData(reinterpret_cast<char *>(ptr), static_cast<int>(len));
    return bytes;
}

QString Response::text() const {
    QByteArray bytes = this->bytes();
    return QString::fromUtf8(bytes);
}

QJsonDocument Response::json() const {
    QByteArray bytes = this->bytes();
    return QJsonDocument::fromJson(bytes);
}

QString Response::getHeader(const QString &name) const {
    char *err = nullptr;
    char *value = bindings::Response_get_header(this->raw_ptr, name.toUtf8(), &err);
    util::check_err(err);
    QString qvalue = QString::fromUtf8(value);
    bindings::char_free(value);
    return qvalue;
}

QList<Cookie> Response::cookies() const {
    char *err = nullptr;
    char *cookies_ptr = bindings::Response_cookies(this->raw_ptr, &err);
    util::check_err(err);
    QString cookie = QString::fromUtf8(cookies_ptr);
    bindings::char_free(cookies_ptr);
    QJsonArray cookies = QJsonDocument::fromJson(cookie.toUtf8()).array();
    QList<Cookie> result;
    for (QJsonValueRef ck: cookies) {
        result.append(Cookie(ck.toObject()));
    }
    return result;
}


Response::ChunkRange Response::chunks() const {
    char *err = nullptr;
    uint64_t sid = bindings::Response_sid(raw_ptr, &err);
    util::check_err(err);
    return {this->req_ptr, sid};
}

Response::ChunkIterator::ChunkIterator(bindings::ScReq *req, uint64_t sid, bool hasNext) {
    this->req = req;
    this->sid = sid;
    this->hasNext = hasNext;
    if (this->hasNext) {
        char *err = nullptr;
        this->ptr = bindings::ScReq_recv_stream(this->req, this->sid, &this->size, &err);
        util::check_err(err);
        if (this->ptr == nullptr)this->hasNext = false;
    }

}

QByteArray Response::ChunkIterator::operator*() const {
    return QByteArray::fromRawData(reinterpret_cast<const char *>(ptr), static_cast<int>(this->size));
}

Response::ChunkIterator &Response::ChunkIterator::operator++() {
    char *err = nullptr;
    this->ptr = bindings::ScReq_recv_stream(this->req, this->sid, &this->size, &err);
    util::check_err(err);
    if (this->ptr == nullptr)this->hasNext = false;

    return *this;
}

Response::~Response() {
    if (this->raw_ptr == nullptr) { return; }
    bindings::Response_drop(this->raw_ptr);
    this->raw_ptr = nullptr;
}
